stackrox/stackrox
# ROX-10895: Detect role bindings to cluster roles within namespace RBAC evaluator

## Description

The RBAC evaluator gives the ability to provide an overview of permissions within either a specific cluster or namespace for e.g. service accounts, users, groups.

This is done by inspecting the role bindings within the specific namespace / cluster, resolving the bound role and subject, and the given permissions for the specific subjects.

As a high-level overview of RBAC within Kubernetes, [refer to this documentation](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#rolebinding-and-clusterrolebinding).

Now, role bindings, which are scoped to a specific namespace, have the ability to bind to cluster roles. The role binding won't grant permission on the whole cluster, but apply the permissions defined within the cluster role to subjects within the specific namespace.

Previously, for namespaced RBAC evaluation (i.e. when a user doesn't have an access scope with full cluster access), we skipped role bindings which bind to cluster role bindings.

This has some unwanted drawbacks:
Take the following role binding as an example:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: test-rolebinding
  namespace: default
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
- kind: ServiceAccount
  name: dh-test
  namespace: default 
```
The role binding will grant the `dh-test` service account within the namespace cluster-admin permission, as defined within the cluster role:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: cluster-admin
rules:
- apiGroups:
  - '*'
  resources:
  - '*'
  verbs:
  - '*'
- nonResourceURLs:
  - '*'
  verbs:
  - '*'
```

However, when viewing that specific service account within e.g. the UI / API **with a scope that doesn't have full access to the cluster, e.g. only the default namespace**, he will get the impression that the service account has _no_ permissions available:
![service_account_rolebinding_clusterrole](https://user-images.githubusercontent.com/89904305/205811145-3fe75c58-9788-40f4-9bc6-96cb6a38d9b0.png)

This gives a wrong depiction of the actual RBAC permission state within the specific namespace.

Instead of confusing people with limited access scopes for this particular issue, we will allow them to view permissions of cluster roles, in case they have role bindings which are binding them, irrespective of whether they have the full cluster in their access scope.

This PR will do so, on a conditional basis, and only elevate the context for retrieving _cluster roles_. The overall behavior will stay the same for non-cluster roles.

## Checklist

If any of these don't apply, please comment below.

## Testing Performed

```
1. Create an access scope that is scoped to a single namespace, e.g. `rox-10895`.

2. Create a serviceaccount + rolebinding within the namespace the access scope is scoped to:
apiVersion: v1
kind: ServiceAccount
metadata:
  creationTimestamp: null
  name: dh-test
  namespace: rox-10895
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: rolebinding
  namespace: rox-10895
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
- kind: ServiceAccount
  name: dh-test
  namespace: rox-10895

3. As admin user / user with full cluster scope, check the service account within the UI and observe that the service account has permission to change _everything_ within the namespace (Config Management -> RBAC -> Service Accounts).

4. Switch to a user that has a role which has the previously created access scope.
5. You should observe the same permissions on the service account albeit the specific user doesn't have full access to the cluster (Config Management -> RBAC -> Service Accounts).
```
